src/net/README.md

```diff
@@ -0,0 +1,473 @@
+	dev := wifinina.New(&cfg)
+	...
+}
+```
+
+The Config structure is netdev-specific; consult the netdev package for Config
+details.  In this case, the WiFi credentials are passed.
+
+### Step 2: Hook the netdev into the net package
+
+Tell the net package to use the netdev by calling net.UseNetdev().  Continuing
+with the wifinina example:
+
+```go
+import (
+	"net"
+	"tinygo.org/x/drivers/wifinina"
+)
+
+func main() {
+	cfg := wifinina.Config{Ssid: "foo", Passphrase: "bar"}
+	dev := wifinina.New(&cfg)
+	net.UseNetdev(dev)
+	...
+}
+```
+
+Now, the net package is linked to the netdev so any network I/O will go through
+the netdev.  Calls to net.Dial(), net.Listen() etc will translate to netdev
+socket calls.
+
+The last step is to connect the netdev to an IP network.
+
+### Step 3: Connect to an IP Network
+
+Before the net package is fully functional, connect the netdev to an underlying
+IP network.  For example, a WiFi netdev would connect to a WiFi access point or
+become a WiFi access point; either way, once connected, the netdev has a
+station IP address and is connected on the IP network.
+
+Call dev.NetConnect() to connect the device to an IP network.  Call
+dev.NetDisconnect() to disconnect.  Continuing example:
+
+```go
+import (
+	"net"
+	"tinygo.org/x/drivers/wifinina"
+)
+
+func main() {
+	cfg := wifinina.Config{Ssid: "foo", Passphrase: "bar"}
+	dev := wifinina.New(&cfg)
+	net.UseNetdev(dev)
+
+	dev.NetConnect()
+        
+	// net package calls here
+	
+	dev.NetDisconnect()
+}
+```
+
+Optionally, get notified of IP network connects and disconnects:
+
+```go
+	dev.Notify(func(e net.Event) {
+		switch e {
+		case net.EventNetUp:
+			println("Network UP")
+		case net.EventNetDown:
+			println("Network DOWN")
+	})
+```
+	
+Here is a simple http server listening on port :8080, before and after porting
+from Go "net/http":
+
+#### Before
+```go
+package main
+
+import (
+	"fmt"
+	"net/http"
+)
+
+func main() {
+	http.HandleFunc("/", HelloServer)
+	http.ListenAndServe(":8080", nil)
+}
+
+func HelloServer(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Hello, %s!", r.URL.Path[1:])
+}
+```
+
+#### After
+```go
+package main
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+    
+	"tinygo.org/x/drivers/wifinina"
+)
+
+func main() {
+	cfg := wifinina.Config{Ssid: "foo", Passphrase: "bar"}
+	dev := wifinina.New(&cfg)
+	net.UseNetdev(dev)
+	dev.NetConnect()
+
+	http.HandleFunc("/", HelloServer)
+	http.ListenAndServe(":8080", nil)
+}
+
+func HelloServer(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Hello, %s!", r.URL.Path[1:])
+}
+```
+
+## Using "net/http" Package
+
+TinyGo's net/http package is a partial port of Go's net/http package, providing
+a subset of the full net/http package.  There are a few features excluded
+during the porting process, in particular:
+
+- No HTTP/2 support
+- No TLS support for HTTP servers (no https servers)
+- HTTP client request can't be reused
```

Comment:
Can you explain why replacing net/http is necessary?

Reply:
> I agree we really need to have networking support, but I'm worried the code in this PR is going to be very difficult to maintain. Currently we need a few changes here and there to support a new Go version (especially when that Go version adds new language features), but so far we managed to avoid having to pull code from upstream on each release. This PR would change that in a large way.

There is a maintenance burden, but I didn't see it as difficult.  There are 10 files under src/net, 5 are straight copies, the other 5 are modified.  Someone needs periodically to back-port upstream changes for bug/security fixes.  As part of this PR, I would offer to maintain these files.  Maybe it's just because I've been thru these 10 files so much, it seems easy to maintain.

> In particular, I wonder whether we could keep using the Go stdlib version of packages like net/http and crypto/tls, and if that's not appropriate for microcontrollers maybe have a separate package with lower resources (outside the standard library)? This is especially important for net/http:

tls/crypto: no.  Go's tls/crypto is software support for TLS protocol.  We have mbed TLS offload where the co-controller (e.g. ESP32) handles the TLS protocol.  netdev goes straight to the offload model.

net/http: good luck, I guess is my first thought.  I've tried multiple times to get the full net/http code working without modifications.  Each time I abandon the effort because of too many things are missing in TinyGo.  I should have keep better notes so I could give you details.  I wouldn't mind revisiting if we can make some progress on this PR.  Other packages such as net/url, net/textproto, net/rpc, net/mail, net/smtp just work, as well as the golang.org/x/net (websocket) packages and third party packages (Paho MQTT, etc).

> 1. I expect there are many people who would like to have a fully featured net/http package (with TLS and all that).

There is TLS for net/http, but only on the client side.  So client http.Get/Head/Post/PostForm() calls work for http or https (TLS).  Http server TLS support is not there.  So no http.ListenAndSeverTLS(), but http.ListenAndServe() is OK.

It should be possible to add https server support, if someone needed it.  The mbed TLS offload on these co-controllers we're using can support it.  I'm inclined to wait and see if someone actually needs that support.  I guess it would be cool to run an https on a micro-controller with let's encrypt and plug it directly into the internet?  Are people running internet services on micro-controllers?   My thinking was these embedded devices live on the edge of the internet, not on the interior.  But there is a path forward to move from the edge.

> 2. If we copy+modify upstream net/http that will also put a burden on us to fix any security issues that might arise. I'm not sure we're able to do that.

There are 17 files in net/http, 9 of which are straight copies, and the other 8 are modified.  I can make same offer to maintain these files wrt security and other upstream changes.

Reply:
Sort of talking out of my assumptions here ... so feel free to shoot it down if this is way off base  ... but I wonder it _might_ be feasible use the standard `net` package by linking implementations of syscalls that dispatch to `netdev` implementations?

As an example of what I mean ... NestedVM was a compiler for producing Java binaries from C/C++ code ... in that case, POSIX functionality was integrated using [newlib](https://sourceware.org/newlib/) ... newlib allows users to implement a callback for dispatching syscalls including network ones such as: https://github.com/bgould/nestedvm/blob/master/src/org/ibex/nestedvm/UnixRuntime.java#L174-L185

Not sure if this is feasible ... but if so, I would imagine that the relevant syscall definitions in standard Go should be relatively stable ... almost certainly more stable that the `net` package.

Reply:
While I do see value in merging with upstream Go that may be a herculean task we're asking for. I think we should strive to get this as polished as possible and completely Go-compatible as possible so that in the future we can merge TinyGo net package with upstream net package with not so much as a hiccup. This would mean eliminating any exported function signatures that are not in upstream Go and keeping the Application usage the same, which I believe is identical, if not very similar.

Reply:
> Sort of talking out of my assumptions here ... so feel free to shoot it down if this is way off base ... but I wonder it _might_ be feasible use the standard `net` package by linking implementations of syscalls that dispatch to `netdev` implementations?

This would be the ultimate solution.  Replacing syscall/sockets with netdev/sockets at the portability layer in the net package.  I tried that route and didn't get very far.  I saved the branch in case I wanted to revisit, but if I recall I was bumping into things that were not available in tinygo.  Even the crypto/tls package should port, negating the need to offload TLS, assuming we can do all the crypto required on these controllers.  Let's say all of this is possible, and someone can pull this off in the future.  What would be the migration path from this PR to the ultimate solution?  Let's look at the pieces:

#### Apps
In this PR, apps import "net", "net/http" and "crypto/tls", same as they would if running in normal Go env.  Nothing should change from the apps' perspective when migrating to the ultimate solution.  Users are unaware the rug has been pulled out, upgraded, and slipped back in.

#### Drivers
In this PR, drivers implement the netdev interface, which is basically the BSD socket(2) interface.  The ultimate solution ports the "net" package to netdev rather than syscall/sockets, and there is a good chance nothing changes in the netdev interface during this port, so nothing changes in the drivers.  (BSD sockets will probably be with us forever).  We retain our investment in drivers.

#### Interfaces
There is a loose end with net.UseNetdev() where the app must pick which netdev driver to use.  We don't really have support for network interfaces (eth0, wlan0, etc); we're hard wiring the driver to the app with UseNetdev().  If the ultimate solution can mimic the full OS network interfaces as defined in the "net" package, then maybe UseNetdev moves outside of the apps visibility.  So there would be a change required to the app.  But to support a full OS model of interfaces, there would need to be something equivalent to the "ip" command in Linux to configure and manage the interfaces.  Seems way overkill for embedded.

So I'm feeling pretty good about a migration path forward from this PR.  I'm skeptical the full "net" package could be ported at the syscall level, but just in case it can, I think we have a good migration story.  

If someone wants to take a go at it, this PR would be a good starting point as the netdev driver parts are done.  I started with an empty tinygo/src/net directory and tried to compile one of the drivers/examples/net examples.  Adding/modifying "net" package files as needed to get a clean compile.